Add Block Device Support

will-v-pi opened the pull request:
Adds support for using `ls`, `mkdir`, `cp` (including between host and device), `rm`, and `cat` commands, with LittleFS (used by MicroPython) or FatFS filesystem support.

Works with block devices specified in the binary info, and with partitions. It autodetects the file system if `--filesystem` is not specified, and also supports formatting the block device if required with the `-f` argument.

This fixes #58

will-v-pi committed:
Add bdev commands

Add ls, mkdir, put and get commands to access block devices

Currently only supports littlefs

will-v-pi committed:
Only include bdev if has_libusb

will-v-pi committed:
Fix bazel build

will-v-pi committed:
Merge put and get into cp, add rm, and make ls non-recursive with ability to specify directory

will-v-pi committed:
Add cat command, use partitions option, and format option

will-v-pi committed:
Respect write and format flags on block device

Default to writeable and formattable, but override based on bdev flags, or partition NSBOOT permissions

will-v-pi committed:
Add initial FatFS support

Adds ls and cp commands for FatFS

will-v-pi committed:
Fix Bazel build

will-v-pi committed:
Add filesystem auto-detection

will-v-pi committed:
Add FatFS support to all bdev commands

will-v-pi committed:
Fix support for CircuitPython block device (Fat) and add full Flash Translation Layer (prevents doing lots of erases)

Change Fat sector size back to 512, and consequently add in flash translation error - wasn't needed before with 4096 sector size

Also enable LFN support

will-v-pi committed:
Fix segfault in release builds, and add output to all commands

will-v-pi committed:
Add commands to readme

will-v-pi committed:
Add support for using partition with ID "blockdev"

Need to put BLOCK_DEVICE_PARTITION_ID into the SDK somewhere

lurch commented on README.md:
```diff
@@ -1392,6 +1394,265 @@ OPTIONS:
+    picotool bdev ls <dirname> [-r] [-p <partition>] [--filesystem <fs>] [-f] [device-selection]
+
+OPTIONS:
+        <dirname>
+            The file name
```
"dirname" vs. "file name" ? :wink: 
Also, if I wanted to list the "root directory" would I give it a `<dirname>` of `/`, or something else?

will-v-pi replied:
Just pass no `dirname`, like you would to the normal `ls` command - it’s an optional argument

lurch commented on README.md:
```diff
@@ -1392,6 +1394,265 @@ OPTIONS:
+    picotool bdev mkdir <dirname> [-p <partition>] [--filesystem <fs>] [-f] [device-selection]
+
+OPTIONS:
+        <dirname>
+            The file name
```
```suggestion
        <dirname>
            The directory name
```

lurch commented on README.md:
```diff
@@ -1392,6 +1394,265 @@ OPTIONS:
+            Specify filesystem to use
+        <fs>
+            littlefs|fatfs
+        -f, --format
```
Hmmm, we now seem to have **BOTH** `-f, --format` and `-f, --force` flags?! Either this worn't work properly, or if it does work it's pretty confusing!

will-v-pi replied:
Good spot - I’ll make this filesystem one not have a short argument

lurch replied:
(perhaps the arg-parsing library could/should raise an error if given the same argument multiple times?)

will-v-pi replied:
> (perhaps the arg-parsing library could/should raise an error if given the same argument multiple times?)

I will leave that for a future PR

lurch commented on README.md:
```diff
@@ -1392,6 +1394,265 @@ OPTIONS:
+
+### cp
+
+Copy file to/from the block device - use :filename to indicate files on the device (eg `cp main.py :main.py` to upload to the device)
```
Probably doesn't hurt to be explicit?
```suggestion
Copy file to/from the block device - use :filename to indicate files on the device (e.g. `cp main.py :main.py` to upload to the device, or `cp :main.py main.py` to download from the device)
```

lurch commented on README.md:
```diff
@@ -1392,6 +1394,265 @@ OPTIONS:
+
+### cp
+
+Copy file to/from the block device - use :filename to indicate files on the device (eg `cp main.py :main.py` to upload to the device)
```
* Does this `bdev cp` command check that one (and only one) of either `src` or `dest` starts with a colon?
* Does it only work for files and not directories?
* If the answer to the previous question is "only files", would it be a lot of additional work to be able to copy directories? (which would presumably need to operate recursively)

will-v-pi replied:
- No - you can copy a file to somewhere else on the device, or even use this to copy local files if you really wanted
- Yes, it only works for files
- Could probably add that - just needs a recursive walk like `ls` has

will-v-pi replied:
Coming back to recursive copying, I think that is trickier, as it needs a recursive walk on the host side - I think I won't add that in this PR, but maybe a future `-r` argument could be added

lurch commented on lib/littlefs/README.md:
```diff
@@ -0,0 +1,7 @@
+The upstream source for the files in this directory is
+https://github.com/littlefs-project/littlefs
+
+The current files come from v2.8.2
```
_Maybe_ worth bumping (after testing, of course) to the latest release? https://github.com/littlefs-project/littlefs/releases/tag/v2.11.1

will-v-pi replied:
Bumped, and seems to work fine

lurch commented on main.cpp:
```diff
@@ -786,6 +829,115 @@ struct config_command : public cmd {
+
+    group get_cli() override {
+        return (
+            optional_untyped_file_selection_x("dirname", 0) +
```
Does the `optional_` here indicate that this is an optional argument? If so, it should probably appear in the `README.md` as `[dirname]` rather than `<dirname>` ?

will-v-pi replied:
Yes, it’s optional - I’m not sure why the CLI stuff isn’t putting square brackets around it

will-v-pi replied:
Fixed

lurch commented on main.cpp:
```diff
@@ -2205,34 +2360,45 @@ struct picoboot_memory_access : public memory_access {
+                }
+                // Check if we need to erase (ie check for non 0xff)
+                if (!std::all_of(write_data.cbegin(), write_data.cend(), [](uint8_t v) { return v == 0xff; })) {
+                    // Do automatically erase flash, and make it aligned
```
```suggestion
                    // To automatically erase flash, and make it aligned
```
?

will-v-pi replied:
The comments were meant to mean “Check if we want to erase” followed by “Yes, we do want erase” - so maybe just delete the “Do”?
```suggestion
                    // Automatically erase flash, and make it aligned
```

lurch replied:
Ahhh, I didn't realise that
```c
// Do automatically erase flash, and make it aligned
// we have to erase in whole pages
```
was intended to be two separate sentences, and not one long sentence :slightly_smiling_face:

lurch commented on main.cpp:
```diff
@@ -2205,34 +2360,45 @@ struct picoboot_memory_access : public memory_access {
+                if (std::equal(write_data.cbegin(), write_data.cend(), buffer)) {
+                    return;
+                }
+                // Check if we need to erase (ie check for non 0xff)
```
AIUI, when you erase Flash it sets it to all-1s, and so when you "write data" to Flash, it only flips some of the 1s to 0s. So I guess that _theoretically_, if you're overwriting `0b11111110` with `0b11111100` then there's also no need to erase the Flash first? Which suggests we could do something more sophisticated here than just "check for non 0xff" ? :thinking:

will-v-pi replied:
That’s true, I could probably check for `write_data & ~current_data` instead

lurch commented on main.cpp:
```diff
@@ -5913,6 +6090,738 @@ bool link_command::execute(device_map &devices) {
+        settings.filenames[1].erase(0, 1);
+    }
+
+    fos << "Copying " << (srcRemote ? "device " : "local ") << settings.filenames[0] << " to " << (destRemote ? "device " : "local ") << settings.filenames[1] << "\n";
```
Perhaps it's worth checking that `(srcRemote || destRemote)`, as there seems to be little point using `picotool bdev cp` to copy one local file to another local file? :wink:

will-v-pi replied:
I don’t think there’s any point removing that functionality, in case there’s a use case where you just use `picotool bdev cp` for everything - and this message makes it clear what’s happening when you do that, so if you didn’t mean to you can add a colon

lurch commented on main.cpp:
```diff
@@ -5913,6 +6090,738 @@ bool link_command::execute(device_map &devices) {
+        auto infile = get_file_idx(ios::in|ios::binary|ios::ate, 0);
+        auto size = infile->tellg();
+        infile->seekg(0, std::ios::beg);
+        data_buf.resize(size);
```
If you accidentally point this at the wrong local filename (e.g. a huge disk-image file), is there a risk that this might cause an out-of-memory error?

will-v-pi replied:
Yes - should probably check the input file fits in the largest flash size (32MB currently)

lurch replied:
That raises an interesting point - are you able to calculate how much "free space" there is in the block device, and what happens if you try writing a file larger than this? Hopefully it doesn't just continue writing to Flash after the end of the block device?

will-v-pi replied:
That's up to the filesystem - both LittleFS and FatFS should throw an error if there isn't enough free space on the device, but I might add a backstop to prevent writing past the end of the block device too

lurch commented on README.md:
```diff
@@ -1392,6 +1394,265 @@ OPTIONS:
+
+### cat
+
+Print contents of file on the block device
```
Will an appropriate error get displayed if you try to `bdev cat` a directory name?

will-v-pi replied:
I've pushed a change to print error strings instead of numbers, so you'd get `ERROR: LittleFS Open Error: Entry is a dir`

lurch commented:
A thought I just had - to make this slightly easier to use, could you have `picotool` check what filename it was invoked with (e.g. like `busybox` does), and then you could have `picotool_ls`, `picotool_cat`, etc. symlinks to the `picotool` binary, and then calling `picotool_ls somedir` would do the same thing as `picotool bdev ls somedir` ? (Or if that's not possible, perhaps the picotool CMake install could add a bunch of shell scripts e.g. named `picotool_ls` which just does `picotool bdev ls $@`)

will-v-pi commented:
> A thought I just had - to make this slightly easier to use, could you have `picotool` check what filename it was invoked with (e.g. like `busybox` does), and then you could have `picotool_ls`, `picotool_cat`, etc. symlinks to the `picotool` binary, and then calling `picotool_ls somedir` would do the same thing as `picotool bdev ls somedir` ? (Or if that's not possible, perhaps the picotool CMake install could add a bunch of shell scripts e.g. named `picotool_ls` which just does `picotool bdev ls $@`)

I think that might be best left to the user to do, to avoid polluting their shell (eg you might want `picotool_ls` to list devices, not use the `bdev` command) - you could also create an alias to `picotool bdev` in bashrc (which also avoids the need for underscores and therefore the shift key), so I don’t think I’ll include that is this PR

will-v-pi committed:
Review fixups

Add extra checks to prevent too big files

Fix dirname description

Remove short format option

Improve erase check (rather than just checking for FF)

will-v-pi committed:
Update readme

will-v-pi committed:
Fix mkdir dirname desc

will-v-pi committed:
Add error strings

will-v-pi committed:
Update LittleFS to v2.11.1

lurch commented on main.cpp:
```diff
@@ -6613,6 +6647,11 @@ bool bdev_cp_command::execute(device_map &devices) {
         auto size = infile->tellg();
+        uint32_t max_size = bdevfs_setup.access->get_model()->flash_end() - bdevfs_setup.access->get_model()->flash_start();
+        if (size > max_size) {
+            fail(ERROR_NOT_POSSIBLE, "File size %dMB is too large for flash device (max %dMB)", size / (1024 * 1024), max_size / (1024 * 1024));
```
Is it possible for this to (also?) check against the "block device" size, rather than just the "flash device" size? Or would that be a premature optimisation because it'll get caught by later checks anyway?

will-v-pi replied:
That'll get caught by later checks by the filesystem anyway - this check is just to prevent reading a multi-gigabyte file into memory

lurch commented on README.md:
```diff
@@ -1392,6 +1394,266 @@ OPTIONS:
+            Partition to use as block device
+        <partition>
+            partition number
+        --filesystem
```
Is the `--filesystem` flag only needed when formatting a block-device, or are there likely to be circumstances in which the filesystem auto-deteciton doesn't work?

will-v-pi replied:
With the current supported filesystems it should only be needed when formatting, but future filesystems may be harder to auto-detect so may require this flag

lurch commented on README.md:
```diff
@@ -1392,6 +1394,266 @@ OPTIONS:
+            Specify filesystem to use
+        <fs>
+            littlefs|fatfs
+        --format
+            Format the drive if necessary (may result in data loss)
```
Presumably using the `--format` option in conjunction with `picotool bdev ls` will **always** return an empty directory listing? :joy:  But this makes me wonder why `--format` is an option that can be used in combination  with any other `picotool bdev` command, rather than just being a standalone `picotool bdev format` or `picotool format_bdev` command? :thinking:

will-v-pi replied:
I added it as a flag because of where the formatting occurs in the code - it will only format the drive if no filesystem (of the type specified by `--filesystem` if that is passed) is detected, so you could always call your first command with `--format` if you wanted to make sure you always do the command. Eg this works for both formatting new devices and updating the file on old devices:
```
picotool bdev cp main.py :/ --format --filesystem littlefs
```

lurch replied:
Oh! I see now that I'd misread "Format the drive if necessary" as just "Format the drive" :facepalm: 

So I guess that means if you had a blockdevice that was already fatfs-formatted, and you wanted to explicitly reformat it, you'd need to do something like:
```
picotool bdev ls --format --filesystem littlefs
picotool bdev ls --format --filesystem fatfs
```
?

will-v-pi replied:
Yes, so it probably is worth adding a separate `picotool bdev format` command to unconditionally format the block device

will-v-pi replied:
Added

lurch commented on README.md:
````diff
@@ -1392,6 +1394,266 @@ OPTIONS:
+Print contents of file on the block device
+
+```text
+$ picotool help bdev cat
+BDEV CAT:
+    Print contents of file on the block device
````
So I guess `picotool bdev cat main.py > main.py` effectively does the same thing as `picotool bdev cp :main.py main.py` ?

will-v-pi replied:
Yes - this also allows for things like `picotool bdev cat main.py | sha256sum`

lurch commented on main.cpp:
```diff
@@ -94,6 +97,10 @@ static __forceinline int __builtin_ctz(unsigned x) {
 #endif
+#ifndef BLOCK_DEVICE_PARTITION_ID
+// The default 0x626C6F636B646576 value is the ASCII encoding of "blockdev"
+#define BLOCK_DEVICE_PARTITION_ID 0x626C6F636B646576
```
Should this ID also appear somewhere in a pico-sdk header, in case 3rd-party utilities also want to use it? Or do we expect any 3rd-party utilities to just copy'n'paste this into their own source code?

will-v-pi replied:
Yes, if it's decided that this is wanted as a way to identify block device partitions, then I'll open an SDK PR to add this to the SDK

kilograham replied:
sounds good

will-v-pi replied:
Done - https://github.com/raspberrypi/pico-sdk/pull/3015

lurch commented on main.cpp:
```diff
@@ -2205,34 +2364,52 @@ struct picoboot_memory_access : public memory_access {
+                if (std::equal(write_data.cbegin(), write_data.cend(), buffer)) {
+                    return;
+                }
+                // Check if we need to erase (ie check for non 0xff)
```
Does this comment need to be updated?

lurch commented on main.cpp:
```diff
@@ -5913,6 +6101,820 @@ bool link_command::execute(device_map &devices) {
+        bdevfs_setup.base_addr = start;
+        bdevfs_setup.size = end - start;
+        block_device_found = true;
+    } else if (partitions) { // use first partition with id "blockdev"
```
It's probably a bit of an extreme edge-case, but does there need to be a mechanism for ignoring a partition named "blockdev", if you happen to have a partition with that name, but you still want to use the block device specified in the binary info?

will-v-pi replied:
It's not a partition named `blockdev` (the partition name can be anything), it's a partition with the 64-bit ID of the ascii characters `blockdev`, which I think is an intentional enough thing to do that we don't need to add handling for ignoring it?

lurch commented on main.cpp:
```diff
@@ -5913,6 +6101,820 @@ bool link_command::execute(device_map &devices) {
+        }
+
+        // No FS Found
+        fail(ERROR_CONNECTION, "No file system detected - to format the drive use `-f --filesystem <littlefs|fatfs>`");
```
I guess the `-f` here needs to be changed to `--format`?

will-v-pi committed:
fixups

will-v-pi committed:
Merge remote-tracking branch 'origin/develop' into block-device-support

will-v-pi committed:
Fixups from develop merge

will-v-pi committed:
Fix bazel build

will-v-pi committed:
Hopefully fix Windows build

kilograham commented on main.cpp:
```diff
@@ -6081,6 +6269,820 @@ bool link_command::execute(device_map &devices) {
+        block_device_found = true;
+    } else if (partitions) { // use first partition with id "blockdev"
+        for (auto &partition : *partitions) {
+            if (std::get<3>(partition) != BLOCK_DEVICE_PARTITION_ID) {
```
seems a bit mean to not let them specify the partition id (or name)

will-v-pi replied:
Added - you can now pass the partition number, name or ID (prioritised in that order)

kilograham commented on main.cpp:
```diff
@@ -6081,6 +6269,820 @@ bool link_command::execute(device_map &devices) {
+                std::stringstream ss;
+                ss << hex_string(bi_bdev.address) << "-" << hex_string(bi_bdev.address + bi_bdev.size) <<
+                    " (" << ((bi_bdev.size + 1023) / 1024) << "K): " << read_string(access, bi_bdev.name);
+                if (bi_bdev.flags) {
```
ditto can we let them select

will-v-pi replied:
Added options to override the flags - `--force-formattable` and `--force-writeable`

kilograham commented on README.md:
```diff
@@ -1660,6 +1662,306 @@ OPTIONS:
 
+## bdev
+
+The `bdev` commands are for interacting with block devices in Flash. These can either be defined using binary info, or you can specify a partition to use as a block device. They work with a LittleFS filesystem, or a FatFS filesystem.
```
These can be defined is confusing; who is defining what, when? perhaps you mean selected not defined. Also perhaps "the commands" rather than They

will-v-pi replied:
```suggestion
The `bdev` commands are for interacting with block devices in Flash. The block device location can either be determined using binary info, or you can specify a partition to use as a block device. The commands work with a LittleFS filesystem, or a FatFS filesystem.
```

kilograham commented on README.md:
```diff
@@ -1660,6 +1662,306 @@ OPTIONS:
+            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
+            command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
+            USB drive mounted
+        --bootsel-led <gpio>
```
probably a minor grumble, and maybe these are mostly applicable to all commands as you can force a reboot, and therefore there can be an activity LED, so maybe its fine... its just  abit weird that it is in target defice selection

will-v-pi replied:
Yes, these are applicable to all commands that can force a reboot. Maybe the device selection could be refactored to show cleaner help text as it is the same for all commands (eg a `picotool help device_selection` command to list this help, rather than listing it out for all commands), but that would be for a different PR to do.

kilograham reviewed (changes requested):
might be worth adding some tests to make sure none of the commands delete everything ;-)

will-v-pi committed:
BLOCK_DEVICE_PARTITION_ID moved into SDK, so just kept here for SDK 2.1.0 and 2.1.1

will-v-pi committed:
Allow picking partition by name or ID

Also rework partition table reading, to use a common function for partition_info and other commands

will-v-pi committed:
review fixup

will-v-pi committed:
Update readme

will-v-pi committed:
Allow overwriting format and write flags

will-v-pi committed:
BLOCK_DEVICE_PARTITION_ID -> BLOCK_DEVICE_DEFAULT_PARTITION_ID

lurch commented on README.md:
```diff
@@ -1703,7 +1703,7 @@ BDEV LS:
 SYNOPSIS:
     picotool bdev ls <dirname> [-r] [-p <partition number>] [--partition-name <partition name>] [--partition-id <partition id>]
-                [--filesystem <fs>] [--format] [device-selection]
+                [--filesystem <fs>] [--format] [--force-formattable] [--force-writeable] [device-selection]
```
Does the `bdev ls` command really need a `--force-writeable` command, given that `ls` will presumably never result in any writes?

will-v-pi replied:
Technically yes, in order to allow formatting a read-only drive - not useful, but technically valid

lurch replied:
Huh, so merely specifying `--force-formattable` wouldn't allow a read-only drive to be formatted? That might be surprising behaviour? :thinking:

will-v-pi replied:
It throws a reasonable error - I think it's fine, as it requires you to stop and think even more about whether you should format the drive.
```
$ picotool bdev ls --partition-id 0 --format --force-formattable --filesystem littlefs
/
Formatting LittleFS file system
ERROR: This block device is not writeable
```

In this case, if you then pass `--force-writeable` too you still get an error, because the partition is only `NSBOOT(R)`
```
$ picotool bdev ls --partition-id 0 --format --force-formattable --force-writeable --filesystem littlefs
/
Formatting LittleFS file system
ERROR: The RP2350 device returned an error: permission failure
```

lurch replied:
Hmmm, now that there's an explicit `bdev format` command (hooray), _maybe_ it doesn't make sense for the `bdev ls` command to have a `--format` flag, because presumably a freshly-formatted partition will always be empty? :thinking:

will-v-pi replied:
It has uses - you might have a provisioning system that checks for the existence of a specifc file on a board, but which needs to deal with both formatted and un-formatted boards, and is expected to output formatted boards. You could use the same command for all of them because all `bdev` commands come with the ability to format first:
```
picotool bdev ls --format --filesystem littlefs
```

lurch replied:
That seems a bit convoluted, but fair enough :slightly_smiling_face: 
If somebody wanted to do things "the long way round" i.e. avoiding any risk of data-loss, is there a `picotool bdev` command to check if a block device is correctly formatted or not?

will-v-pi replied:
Yes, they can use any command except format, and the command will throw an error if the device wasn’t correctly formatted. If you pass `--filesystem` as well, it will throw an error if the device isn’t formatted to the specified file system.

lurch replied:
Cool, thanks. Is that behaviour worth documenting (i.e. how to check if a block device is formatted, without writing to it), or do you think that it's sufficiently obvious that people will figure it out for themselves?

will-v-pi replied:
I think it's clear enough?
```
$ picotool bdev ls
ERROR: No file system detected - to format the drive use `--format --filesystem <littlefs|fatfs>`

$ picotool bdev ls --filesystem littlefs
/
ERROR: LittleFS file system is corrupted - add --format flag to format it (this may result in data loss)

$ picotool bdev ls --filesystem fatfs
/
ERROR: FatFS file system is corrupted - add --format flag to format it (this may result in data loss)
```

will-v-pi committed:
Add format command

will-v-pi committed:
Fix optional_untyped_path_selection_x and optional_untyped_file_selection_x so they are documented as optional

will-v-pi committed:
Allow passing : filenames to other bdev commands

This matches the cp command where : filenames are on-device, but still allow non : filenames as these commands only operate on-device

lurch commented on main.cpp:
```diff
@@ -732,12 +781,14 @@ auto device_selection =
-        }).set(settings.filenames[i]).min(0) % "The file name"\
-).min(0).doc_non_optional(true)
+        }).set(settings.filenames[i]).min(0) % desc\
+).min(0)
```
Do the other `opitional_...` macros (e.g. `optional_file_selection_x` and `optional_typed_file_selection_x`) also need the `.doc_non_optional(true)` to be removed?

will-v-pi replied:
No, it’s only this one - the others all have 2 options inside the brackets (file name and type), so seem to need the `doc_non_optional` to be displayed correctly - but this one only has 1 option inside the brackets, so is displayed as optional already. I’m not sure why that’s the case though…

lurch replied:
Fair enough - I guess this is one of those "if it works, don't touch it" things :rofl:

will-v-pi committed:
Merge remote-tracking branch 'origin/develop' into block-device-support

will-v-pi committed:
Suppress compile warning from oofatfs

will-v-pi committed:
Only suppress warnings when not on Windows

will-v-pi committed:
Fix corrupt fatfs error message

will-v-pi committed:
Add bdev_test.sh script

Runs some basic tests of adding and removing files

aswild commented on main.cpp:
```diff
@@ -4354,35 +4579,66 @@ std::shared_ptr<vector<tuple<uint32_t, uint32_t>>> get_partitions(picoboot::conn
+    return std::make_shared<tuple<resident_partition_t, std::shared_ptr<vector<partition_details>>>>(std::make_tuple(unpartitioned, std::make_shared<vector<partition_details>>(ret)));
+}
+
+std::shared_ptr<vector<partition_details>> get_partitions(picoboot::connection &con) {
+    return std::get<1>(*get_partition_info(con));
```
this operation crashes `picotool info` on RP2040 targets that don't support partition info. When `model->supports_partition_table()` returns `false`, as it does on RP2040, `get_partition_info()` returns `nullptr` and then this deref fails (either with a failed assert in debug mode, or a segfault in release mode).

This code needs to forward on `nullptr` (everywhere that calls `get_partitions` already makes sure its result is non-null before using it)

```diff
diff --git a/main.cpp b/main.cpp
index 0c3b093..b2d37e4 100644
--- a/main.cpp
+++ b/main.cpp
@@ -4670,7 +4670,11 @@ std::shared_ptr<tuple<resident_partition_t, std::shared_ptr<vector<partition_det
 }

 std::shared_ptr<vector<partition_details>> get_partitions(picoboot::connection &con) {
-    return std::get<1>(*get_partition_info(con));
+    auto info = get_partition_info(con);
+    if (info) {
+        return std::get<1>(*info);
+    }
+    return nullptr;
 }
 #endif

```

lurch replied:
> this operation crashes `picotool info` on RP2040 targets that don't support partition info.

Thanks for the report. Could you create a new issue please?
